README.md
# kotlin-livedata-utils

 [![Download](https://api.bintray.com/packages/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/images/download.svg) ](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/_latestVersion)
 [![GitHub tag](https://img.shields.io/github/tag/magneticflux-/kotlin-livedata-utils.svg)](https://github.com/magneticflux-/kotlin-livedata-utils/tags) 
[![Build Status](https://travis-ci.com/magneticflux-/kotlin-livedata-utils.svg?branch=master)](https://travis-ci.com/magneticflux-/kotlin-livedata-utils) 
[![GitHub issues](https://img.shields.io/github/issues/magneticflux-/kotlin-livedata-utils.svg)](https://github.com/magneticflux-/kotlin-livedata-utils/issues) 
[![GitHub PRs](https://img.shields.io/github/issues-pr/magneticflux-/kotlin-livedata-utils.svg)](https://github.com/magneticflux-/kotlin-livedata-utils/pulls) 
[![GitHub Contributors](https://img.shields.io/github/contributors/magneticflux-/kotlin-livedata-utils.svg)](https://github.com/magneticflux-/kotlin-livedata-utils/graphs/contributors) 
[![GitHub license](https://img.shields.io/github/license/magneticflux-/kotlin-livedata-utils.svg)](https://github.com/magneticflux-/kotlin-livedata-utils/blob/master/LICENSE) 

[![Get automatic notifications about new "kotlin-livedata-utils" versions](https://www.bintray.com/docs/images/bintray_badge_color.png)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils?source=watch)

### Usage:

```Gradle
repositories {
    jcenter()
}

dependencies {
    implementation 'com.github.magneticflux:kotlin-livedata-utils:$latest_version'
}
```

```Kotlin
import com.github.magneticflux.livedata.*

// ...
{
    // ...
    return uuid.switchMap { itemRepository.getItemByUUID(it) }
}
```

### Compatibility:

Library Version       | Android Support Library                 | Android API Level
----------------------|-----------------------------------------|------------------
ver ∈ \[0.1.0, 0.2.X] | `android.arch.lifecycle:livedata`       | ver ≥ 14
ver ∈ \[0.3.0, 0.3.X] | `androidx.lifecycle:lifecycle-livedata` | ver ≥ 14

---
Copyright 2018, Mitchell Skaggs

Licensed under the GNU Lesser General Public License v3.0


docs/release-notes.md
<sup><sup>*Release notes were automatically generated by [Shipkit](http://shipkit.org/)*</sup></sup>

#### 0.3.4
 - 2019-07-30 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.3.3...v0.3.4) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.3.4-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.3.4)
 - No pull requests referenced in commit messages.

#### 0.3.3
 - 2018-07-06 - [4 commits](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.3.2...v0.3.3) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.3.3-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.3.3)
 - Feature/dependency updates [(#7)](https://github.com/magneticflux-/kotlin-livedata-utils/pull/7)

#### 0.3.2
 - 2018-06-25 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.3.1...v0.3.2) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.3.2-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.3.2)
 - No pull requests referenced in commit messages.

#### 0.3.1
 - 2018-06-25 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.3.0...v0.3.1) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.3.1-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.3.1)
 - No pull requests referenced in commit messages.

#### 0.3.0
 - 2018-06-25 - [7 commits](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.2.0...v0.3.0) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.3.0-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.3.0)
 - Feature/androidx updates [(#4)](https://github.com/magneticflux-/kotlin-livedata-utils/pull/4)

#### 0.2.0
 - 2018-06-08 - [5 commits](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.7...v0.2.0) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.2.0-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.2.0)
 - Added zipLiveDataNullable function [(#3)](https://github.com/magneticflux-/kotlin-livedata-utils/pull/3)

#### 0.1.7
 - 2018-06-08 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.6...v0.1.7) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.7-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.7)
 - No pull requests referenced in commit messages.

#### 0.1.6
 - 2018-06-07 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.5...v0.1.6) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.6-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.6)
 - No pull requests referenced in commit messages.

#### 0.1.5
 - 2018-06-07 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.4...v0.1.5) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.5-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.5)
 - No pull requests referenced in commit messages.

#### 0.1.4
 - 2018-06-06 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.3...v0.1.4) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.4-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.4)
 - No pull requests referenced in commit messages.

#### 0.1.3
 - 2018-06-06 - no code changes (no commits) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.3-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.3)

#### 0.1.2
 - 2018-06-06 - [1 commit](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.1...v0.1.2) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.2-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.2)
 - No pull requests referenced in commit messages.

#### 0.1.1
 - 2018-06-06 - [2 commits](https://github.com/magneticflux-/kotlin-livedata-utils/compare/v0.1.0...v0.1.1) by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.1-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.1)
 - No pull requests referenced in commit messages.

#### 0.1.0
 - 2018-06-06 - 5 commits by [Mitchell Skaggs](https://github.com/magneticflux-) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.1.0-green.svg)](https://bintray.com/magneticflux/kotlin-livedata-utils/kotlin-livedata-utils/0.1.0)
 - No pull requests referenced in commit messages.



kotlin-livedata-utils/src/test/java/com/github/magneticflux/livedata/LiveDataUnitTest.kt
package com.github.magneticflux.livedata

import android.arch.lifecycle.MediatorLiveData
import android.arch.lifecycle.MutableLiveData
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith


@ExtendWith(InstantTaskExecutorExtension::class)
class LiveDataUnitTest {

    @Test
    fun `when two values are zipped together with zipTo, they ignore nulls`() {
        val v1 = MutableLiveData<String>()
        val v2 = MutableLiveData<String>()

        val pair = v1 zipTo v2

        v1.value = null
        v2.value = null
        Assertions.assertEquals(null, pair.awaitValue)

        v1.value = "Hello"
        v2.value = "world"
        Assertions.assertEquals("Hello" to "world", pair.awaitValue)

        v1.value = "Goodbye"
        Assertions.assertEquals("Goodbye" to "world", pair.awaitValue)
    }

    @Test
    fun `when two values are zipped together with zipToNullable, they encounter nulls`() {
        val v1 = MutableLiveData<String>()
        val v2 = MutableLiveData<String>()

        val pair = v1 zipToNullable v2

        v1.value = null
        v2.value = null
        Assertions.assertEquals(null to null, pair.awaitValue)

        v1.value = "Hello"
        v2.value = "world"
        Assertions.assertEquals("Hello" to "world", pair.awaitValue)

        v1.value = "Goodbye"
        Assertions.assertEquals("Goodbye" to "world", pair.awaitValue)
    }

    @Test
    fun `when a null value is passed to a map, null is returned`() {
        val nullable = MutableLiveData<String>()
        val mapped = nullable.map { "$it world!" }

        nullable.value = null
        Assertions.assertNull(mapped.awaitValue)

        nullable.value = "Hello"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a null value is passed to a mapNullable, null is encountered`() {
        val nullable = MutableLiveData<String>()
        val mapped = nullable.mapNullable { "$it world!" }

        nullable.value = null
        Assertions.assertEquals("null world!", mapped.awaitValue)

        nullable.value = "Hello"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a null value is passed to a switchMap, null is returned`() {
        val nullable = MutableLiveData<String>()

        val data = MutableLiveData<String>().apply { value = "Hello world!" }
        val backup = MutableLiveData<String>().apply { value = "Error!" }

        val mapped = nullable.switchMap {
            if (it == null)
                backup
            else
                data
        }

        nullable.value = null
        Assertions.assertNull(mapped.awaitValue)

        nullable.value = "Hello world!"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a null value is passed to a switchMapNullable, null is encountered`() {
        val nullable = MutableLiveData<String>()

        val data = MutableLiveData<String>().apply { value = "Hello world!" }
        val backup = MutableLiveData<String>().apply { value = "Error!" }

        val mapped = nullable.switchMapNullable {
            if (it == null)
                backup
            else
                data
        }

        nullable.value = null
        Assertions.assertEquals("Error!", mapped.awaitValue)

        nullable.value = "Hello world!"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a bidirectional map is established, changes flow in both directions`() {

        val aAndB = MediatorLiveData<Pair<String, String>>()
        val a = aAndB.bidiMap(
                { pair: Pair<String, String>, _: String? ->
                    pair.first
                },
                { oldPair: Pair<String, String>, newString: String ->
                    oldPair.copy(first = newString)
                }
        )
        val b = aAndB.bidiMap(
                { pair: Pair<String, String>, _: String? ->
                    pair.second
                },
                { oldPair: Pair<String, String>, newString: String ->
                    oldPair.copy(second = newString)
                }
        )


        aAndB.value = "abc" to "123"

        Assertions.assertEquals(a.awaitValue, "abc")
        Assertions.assertEquals(b.awaitValue, "123")

        a.value = "xyz"

        Assertions.assertEquals(aAndB.awaitValue, "xyz" to "123")

        b.value = "789"

        Assertions.assertEquals(aAndB.awaitValue, "xyz" to "789")
    }
}

kotlin-livedata-utils/src/test/java/com/github/magneticflux/livedata/ExampleUnitTest.java
package com.github.magneticflux.livedata;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() {
        assertEquals(4, 2 + 2);
    }
}

kotlin-livedata-utils/src/test/java/com/github/magneticflux/livedata/LiveDataTestUtils.kt
package com.github.magneticflux.livedata

import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer

/**
 * Waits a ***very*** short amount of time for a value to be present. Must be used with [InstantTaskExecutorExtension].
 */
val <T> LiveData<T>.awaitValue: T?
    get() {

        lateinit var observer: Observer<T>
        observer = Observer {
            removeObserver(observer)
        }
        this.observeForever(observer)

        return this.value
    }

kotlin-livedata-utils/src/android/java/com/github/magneticflux/livedata/LiveDataUtils.kt
package com.github.magneticflux.livedata

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.Transformations

/**
 * This function creates a [LiveData] of a [Pair] of the two types provided. The resulting LiveData is updated whenever either input LiveData updates and both LiveData have updated to a non-null value at least once before.
 *
 * If the zip of A and B is C, and A and B are updated in this pattern: `AABA`, C would be updated twice (once with the second A value and first B value, and once with the third A value and first B value).
 *
 * @param a the first LiveData
 * @param b the second LiveData
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
fun <A, B> zipLiveData(a: LiveData<A>, b: LiveData<B>): LiveData<Pair<A, B>> {
    return MediatorLiveData<Pair<A, B>>().apply {
        var lastA: A? = null
        var lastB: B? = null

        fun update() {
            val localLastA = lastA
            val localLastB = lastB
            if (localLastA != null && localLastB != null)
                this.value = Pair(localLastA, localLastB)
        }

        addSource(a) {
            lastA = it
            update()
        }
        addSource(b) {
            lastB = it
            update()
        }
    }
}

/**
 * This function creates a [LiveData] of a [Pair] of the two types provided. The resulting LiveData is updated whenever either input LiveData updates and both LiveData have updated at least once before.
 *
 * @see zipLiveData
 * @param a the first LiveData
 * @param b the second LiveData
 * @since 0.2.0
 * @author Mitchell Skaggs
 */
fun <A, B> zipLiveDataNullable(a: LiveData<A>, b: LiveData<B>): LiveData<Pair<A?, B?>> {
    return MediatorLiveData<Pair<A?, B?>>().apply {
        var lastA: A? = null
        var hasAUpdatedOnce = false

        var lastB: B? = null
        var hasBUpdatedOnce = false

        fun update() {
            if (hasAUpdatedOnce && hasBUpdatedOnce)
                this.value = Pair(lastA, lastB)
        }

        addSource(a) {
            lastA = it
            hasAUpdatedOnce = true
            update()
        }
        addSource(b) {
            lastB = it
            hasBUpdatedOnce = true
            update()
        }
    }
}

/**
 * This is merely an extension function for [zipLiveData].
 *
 * @see zipLiveData
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
infix fun <A, B> LiveData<A>.zipTo(b: LiveData<B>): LiveData<Pair<A, B>> = zipLiveData(this, b)

/**
 * This is merely an extension function for [zipLiveDataNullable].
 *
 * @see zipLiveData
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
infix fun <A, B> LiveData<A>.zipToNullable(b: LiveData<B>): LiveData<Pair<A?, B?>> = zipLiveDataNullable(this, b)

/**
 * This is an extension function that calls to [Transformations.map]. If null is received, null is returned instead of calling the provided function.
 *
 * @see Transformations.map
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
inline fun <A, B> LiveData<A>.map(crossinline function: (A) -> B): LiveData<B> =
        Transformations.map(this) { it: A? ->
            if (it == null) null else function(it)
        }

/**
 * This is an extension function that calls to [Transformations.map]. It exposes the possibilities of receiving and returning null.
 *
 * @see Transformations.map
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
fun <A, B> LiveData<A>.mapNullable(function: (A?) -> B?): LiveData<B> =
        Transformations.map(this, function)

/**
 * This is an extension function that calls to [Transformations.switchMap]. If null is received, null is returned instead of calling the provided function.
 *
 * @see Transformations.switchMap
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
fun <A, B> LiveData<A>.switchMap(function: (A) -> LiveData<B>): LiveData<B> =
        Transformations.switchMap(this) {
            if (it == null) null else function(it)
        }

/**
 * This is an extension function that calls to [Transformations.switchMap]. It exposes the possibilities of receiving and returning null.
 *
 * @see Transformations.switchMap
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
fun <A, B> LiveData<A>.switchMapNullable(function: (A?) -> LiveData<B>?): LiveData<B> =
        Transformations.switchMap(this, function)

/**
 * This is an extension function that links two [MediatorLiveData] instances using a function [apply] and its inverse, [unapply]. Changes made to either [MediatorLiveData] are reflected in the other based on the two functions.
 *
 * @param apply A function from [A] to [B]
 * @param unapply A function from [B] to [A], the inverse of [apply]
 * @since 0.1.0
 * @author Mitchell Skaggs
 */
fun <A, B> MediatorLiveData<A>.bidiMap(apply: (currentA: A, oldB: B?) -> B?, unapply: (oldA: A, currentB: B) -> A?): MediatorLiveData<B> {
    return MediatorLiveData<B>().apply {
        this@apply.addSource(this@bidiMap) {
            it?.let { a ->
                this@apply.value.let { b ->
                    val newB = apply(a, b)
                    if (b != newB)
                        this@apply.value = newB
                }
            }
        }

        this@bidiMap.addSource(this@apply) {
            it?.let { b ->
                this@bidiMap.value?.let { a ->
                    val newA = unapply(a, b)
                    if (a != newA)
                        this@bidiMap.value = newA
                }
            }
        }
    }
}


kotlin-livedata-utils/src/testAndroidx/java/com/github/magneticflux/livedata/LiveDataUnitTest.kt
package com.github.magneticflux.livedata

import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith


@ExtendWith(InstantTaskExecutorExtension::class)
class LiveDataUnitTest {

    @Test
    fun `when two values are zipped together with zipTo, they ignore nulls`() {
        val v1 = MutableLiveData<String>()
        val v2 = MutableLiveData<String>()

        val pair = v1 zipTo v2

        v1.value = null
        v2.value = null
        Assertions.assertEquals(null, pair.awaitValue)

        v1.value = "Hello"
        v2.value = "world"
        Assertions.assertEquals("Hello" to "world", pair.awaitValue)

        v1.value = "Goodbye"
        Assertions.assertEquals("Goodbye" to "world", pair.awaitValue)
    }

    @Test
    fun `when two values are zipped together with zipToNullable, they encounter nulls`() {
        val v1 = MutableLiveData<String>()
        val v2 = MutableLiveData<String>()

        val pair = v1 zipToNullable v2

        v1.value = null
        v2.value = null
        Assertions.assertEquals(null to null, pair.awaitValue)

        v1.value = "Hello"
        v2.value = "world"
        Assertions.assertEquals("Hello" to "world", pair.awaitValue)

        v1.value = "Goodbye"
        Assertions.assertEquals("Goodbye" to "world", pair.awaitValue)
    }

    @Test
    fun `when a null value is passed to a map, null is returned`() {
        val nullable = MutableLiveData<String>()
        val mapped = nullable.map { "$it world!" }

        nullable.value = null
        Assertions.assertNull(mapped.awaitValue)

        nullable.value = "Hello"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a null value is passed to a mapNullable, null is encountered`() {
        val nullable = MutableLiveData<String>()
        val mapped = nullable.mapNullable { "$it world!" }

        nullable.value = null
        Assertions.assertEquals("null world!", mapped.awaitValue)

        nullable.value = "Hello"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a null value is passed to a switchMap, null is returned`() {
        val nullable = MutableLiveData<String>()

        val data = MutableLiveData<String>().apply { value = "Hello world!" }
        val backup = MutableLiveData<String>().apply { value = "Error!" }

        val mapped = nullable.switchMap {
            if (it == null)
                backup
            else
                data
        }

        nullable.value = null
        Assertions.assertNull(mapped.awaitValue)

        nullable.value = "Hello world!"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a null value is passed to a switchMapNullable, null is encountered`() {
        val nullable = MutableLiveData<String>()

        val data = MutableLiveData<String>().apply { value = "Hello world!" }
        val backup = MutableLiveData<String>().apply { value = "Error!" }

        val mapped = nullable.switchMapNullable {
            if (it == null)
                backup
            else
                data
        }

        nullable.value = null
        Assertions.assertEquals("Error!", mapped.awaitValue)

        nullable.value = "Hello world!"
        Assertions.assertEquals("Hello world!", mapped.awaitValue)
    }

    @Test
    fun `when a bidirectional map is established, changes flow in both directions`() {

        val aAndB = MediatorLiveData<Pair<String, String>>()
        val a = aAndB.bidiMap(
                { pair: Pair<String, String>, _: String? ->
                    pair.first
                },
                { oldPair: Pair<String, String>, newString: String ->
                    oldPair.copy(first = newString)
                }
        )
        val b = aAndB.bidiMap(
                { pair: Pair<String, String>, _: String? ->
                    pair.second
                },
                { oldPair: Pair<String, String>, newString: String ->
                    oldPair.copy(second = newString)
                }
        )


        aAndB.value = "abc" to "123"

        Assertions.assertEquals(a.awaitValue, "abc")
        Assertions.assertEquals(b.awaitValue, "123")

        a.value = "xyz"

        Assertions.assertEquals(aAndB.awaitValue, "xyz" to "123")

        b.value = "789"

        Assertions.assertEquals(aAndB.awaitValue, "xyz" to "789")
    }
}